File: smaller_scripts/randomizePlyFilePoints.py
# This script randomizes the order of points in a .ply file:
#     It opens the input PLY file and reads all lines into memory.
#     It divides the lines into header and body, based on the 'end_header' line in the PLY file.
#     It then shuffles the body lines, which correspond to the vertices of your model.
#     Finally, it writes the header and shuffled body lines to a new PLY file.
import random

def shuffle_ply_file(input_filename, output_filename):
    try:
        with open(input_filename, 'r') as f:
            lines = f.readlines()

        header = []
        body = []

        for line in lines:
            if line.startswith('element vertex'):
                num_vertices = int(line.split()[-1])
            if line.startswith('end_header'):
                header.append(line)
                break
            header.append(line)

        body = lines[len(header):]
        body = [body[n:n+1] for n in range(0, len(body), 1)]

        random.shuffle(body)

        with open(output_filename, 'w') as f:
            f.writelines(header)
            for chunk in body:
                f.writelines(chunk)

    except Exception as e: print(e)


shuffle_ply_file('timecube.ply', 'output.ply')

print('Type anything and press enter to exit') 
x = input()
print('Hello, ' + x) 

File: smaller_scripts/randomizePlyFilePoints.js
// This script randomizes the order of points in a .ply file:
//      It opens the input PLY file and reads all lines into memory.
//      It divides the lines into header and body, based on the 'end_header' line in the PLY file.
//      It then shuffles the body lines, which correspond to the vertices of your model.
//      Finally, it writes the header and shuffled body lines to a new PLY file.

import { readFile, writeFile } from 'fs';

function shuffleArray(array) {
    for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]];
    }
    return array;
}

function shufflePlyFile(inputFilename, outputFilename) {
    readFile(inputFilename, 'utf-8', (err, data) => {
        if (err) {
            console.error(err);
            return;
        }
        
        const lines = data.split('\n');
        const headerIndex = lines.indexOf('end_header');
        const headerLines = lines.slice(0, headerIndex + 1);
        const bodyLines = lines.slice(headerIndex + 1, -1);  // -1 to avoid empty line at end of file

        const shuffledBodyLines = shuffleArray(bodyLines);
        const outputData = [...headerLines, ...shuffledBodyLines].join('\n');

        writeFile(outputFilename, outputData, 'utf-8', (err) => {
            if (err) {
                console.error(err);
            } else {
                console.log(`Shuffled PLY file written to ${outputFilename}`);
            }
        });
    });
}

// shufflePlyFile('input.ply', 'output.ply'); //example function call
